src/Numerics/Mesh/Metrics.jl

```diff
@@ -98,127 +100,298 @@ function creategrid!(x1, x2, x3, e2c, ξ1, ξ2, ξ3)
+
+    @inbounds for e in 1:nelem
+        for k in 1:Nq[3], j in 1:Nq[2], i in 1:Nq[1]
+
+            for n in 1:Nq[1]
+                vgeo.x1ξ1[i, j, k, e] += D1[i, n] * vgeo.x1[n, j, k, e]
+                vgeo.x2ξ1[i, j, k, e] += D1[i, n] * vgeo.x2[n, j, k, e]
+                vgeo.x3ξ1[i, j, k, e] += D1[i, n] * vgeo.x3[n, j, k, e]
+            end
+            for n in 1:Nq[2]
+                vgeo.x1ξ2[i, j, k, e] += D2[j, n] * vgeo.x1[i, n, k, e]
+                vgeo.x2ξ2[i, j, k, e] += D2[j, n] * vgeo.x2[i, n, k, e]
+                vgeo.x3ξ2[i, j, k, e] += D2[j, n] * vgeo.x3[i, n, k, e]
+            end
+            for n in 1:Nq[3]
+                vgeo.x1ξ3[i, j, k, e] += D3[k, n] * vgeo.x1[i, j, n, e]
+                vgeo.x2ξ3[i, j, k, e] += D3[k, n] * vgeo.x2[i, j, n, e]
+                vgeo.x3ξ3[i, j, k, e] += D3[k, n] * vgeo.x3[i, j, n, e]
+            end
+
+        end
+    end
+
+    return vgeo
+end
+
+"""
+    computemetric!(vgeo, sgeo, D)
+
+Input arguments:
+- vgeo::VolumeGeometry, a struct containing the volumetric geometric factors
+- sgeo::SurfaceGeometry, a struct containing the surface geometric factors
+- D::DAT2, 1-D derivative operator on the device
 
-Compute the 1-D metric terms from the element grid arrays `x1`. All the arrays
+Compute the 1-D metric terms from the element grid arrays `vgeo.x1`. All the arrays
 are preallocated by the user and the (square) derivative matrix `D` should be
 consistent with the reference grid `ξ1` used in [`creategrid!`](@ref).
 
 If `Nq = size(D, 1)` and `nelem = div(length(x1), Nq)` then the volume arrays
 `x1`, `J`, and `ξ1x1` should all have length `Nq * nelem`.  Similarly, the face
 arrays `sJ` and `n1` should be of length `nface * nelem` with `nface = 2`.
 """
-function computemetric!(x1, J, JcV, ξ1x1, sJ, n1, D)
+function computemetric!(
+    vgeo::VolumeGeometry{NTuple{1,Int}, <:AbstractArray},
+    sgeo::SurfaceGeometry{NTuple{1,Int}, <:AbstractArray},
+    D
+)
+
     Nq = size(D, 1)
-    nelem = div(length(J), Nq)
-    x1 = reshape(x1, (Nq, nelem))
-    J = reshape(J, (Nq, nelem))
-    JcV = reshape(JcV, (Nq, nelem))
-    ξ1x1 = reshape(ξ1x1, (Nq, nelem))
+    nelem = div(length(vgeo.ωJ), Nq)
+
     nface = 2
-    n1 = reshape(n1, (1, nface, nelem))
-    sJ = reshape(sJ, (1, nface, nelem))
+    sgeo.n1 = reshape(sgeo.n1, (1, nface, nelem))
+    sgeo.sωJ = reshape(sgeo.sωJ, (1, nface, nelem))
 
     @inbounds for e in 1:nelem
-        JcV[:, e] = J[:, e] = D * x1[:, e]
+        vgeo.JcV[:, e] = vgeo.ωJ[:, e] = D * vgeo.x1[:, e]
     end
-    ξ1x1 .= 1 ./ J
+    vgeo.ξ1x1 .= 1 ./ vgeo.ωJ
 
-    n1[1, 1, :] .= -sign.(J[1, :])
-    n1[1, 2, :] .= sign.(J[Nq, :])
-    sJ .= 1
+    sgeo.n1[1, 1, :] .= -sign.(vgeo.ωJ[1, :])
+    sgeo.n1[1, 2, :] .= sign.(vgeo.ωJ[Nq, :])
+    sgeo.sωJ .= 1
     nothing
 end
 
 """
-    computemetric!(x1, x2, J, JcV, ξ1x1, ξ2x1, ξ1x2, ξ2x2, sJ, n1, n2, D1, D2)
+    computemetric!(vgeo, sgeo, D1, D2)
```

Comment:
Again, would be good to make this a tuple.
```suggestion
    computemetric!(vgeo, sgeo, (D1, D2))
```

Reply:
Thanks. I went ahead and changed the definition of this function, as well as the `creategrid` one, to accept a tuple as argument, as suggested